Update doc because local doesn't mean local-returning (#2727)
ocaml-flambda/flambda-backend: ocaml/jane/doc/extensions/local/reference.md

> 

Reviewer: The title of this section "local-returning functions" should be renamed to mention `exclave_`. I suggest:

## Use `exclave_` to return a stack allocation

Reviewer: This also requires updating references to this section, e.g. "See "Local-returning functions" below."

Author: Yes I also thought about this, but thought maybe "local-returning functions" is what users want, and "exclave" is just a way (currently the only way) to achieve that. Let me know if you still think we shoud change to "exclave". (I have no preference).

Reviewer: I think that the either the title or the first line of the section should mention “exclave” to help users scan the doc.

maybe the first sentence approach is enough?

> TL;DR: Use `exclave_` to return a local value.

“Local value” is better than “local allocation.”

Author: Thanks - I made the suggested change in #2734

The region around the body of a function prevents local allocations inside that
function from escaping. Occasionally, it is useful to write a function that

> allocates and returns value in caller's region. For instance, consider this code

Reviewer: ```suggestion
allocates and returns a value in caller's region. For instance, consider this code
```

In this code, the counter will *not* be allocated locally. The reason is the
`Counter.make` function: the allocation of `ref 0` escapes the region of
`Counter.make`, and the compiler will therefore not allow it to be locally
allocated. This remains the case no matter how many `local_` annotations we
write inside `f`: the issue is the definition of `make`, not its uses.

To allow the counter to be locally allocated, we need to make `Counter.make`

> allocate and return in caller's region. This can be done by `exclave_`:

Reviewer: ```suggestion
end its region early so that it can allocate its return value in the caller's region. This can be done with `exclave_`:
```

```ocaml
let make () = exclave_
  ref 0
```

The keyword `exclave_` terminates the current region and executes the subsequent
code in the outer region. Therefore, `ref 0` is executed in `f`'s region, which

> allows its local allocation. The allocation will only be cleaned up when the

Reviewer: Note to self: when we merge this PR, we should make a follow-on PR that uses the term "stack allocation" consistently instead of "local allocation".

Author: Yes I think [#2540](https://github.com/ocaml-flambda/flambda-backend/pull/2540#discussion_r1653222006) is very related. My current plan is that that PR will be merged without document and notifying the user. We will (in a few weeks) do a complete overhaul of this document, talk about "stack allocation" and introduce the `stack_` syntax.

region of `f` ends.

## Advanced usage of exclaves
In the previous section, the example function exits its own region immediately,

> which allows allocating and returning in caller's region. This approach,

Reviewer: ```suggestion
which allows allocating and returning in the caller's region. This approach,
```

  else exclave_ Some x
```

In this example, the function `f` has a region where the allocation for the
complex computation occurs. This region is terminated by `exclave_`, releasing
all temporary allocations. Both `None` and `Some x` are considered "local"

> relative to the outer region and are allowed to be returned. In summary, we have
> temporary allocations in the `f`'s region that are promptly released and result
> allocations in the caller's region that can be returned.

Reviewer: ```suggestion
relative to the outer region and are allowed to be returned. In summary, the
temporary allocations in `f`'s region are promptly released, and the result
allocations in the caller's region can be returned.
```

val maybe_length : ('a -> bool) -> 'a list -> local_ int option
```
It is designed not to allocate heap memory by using the stack for all `Some`
allocations. However, it will currently use O(N) stack space because all
allocations occur in the original caller's stack frame. To improve its space

> usage, we delay the `exclave_` annotation until returning result:

Reviewer: ```suggestion
usage, we delay the `exclave_` annotation until returning the result:
```
